torch_xla/csrc/tensor.cpp

```diff
@@ -102,7 +103,14 @@ XLATensor::XLATensor(const at::Tensor& tensor,
 XLATensor::XLATensor(torch::lazy::BackendDataPtr handle,
                      c10::optional<at::ScalarType> logical_element_type)
     : XLATensor(std::make_shared<Data>(handle, handle->device(),
-                                       logical_element_type)) {}
+                                       logical_element_type)) {
+  // if data is sharded we need to carry the sharding spec over.
+  runtime::ComputationClient::DataPtr data = UnwrapXlaData(handle);
+  if (data->HasSharding()) {
+    ShardingSpec sharding_spec(data->GetSharding(), data->shape());
```

Comment:
we need to be consistent I think, I can add a check to not copy over the replicated sharding, but how can we tell if it is implicit replication or manual replication? or do we need to care about the difference between two?

Reply:
I guess we don't? For implicit replication, I believe we attach the replication spec to the data anyway? So for internal or compiler stand point, it's the same. Just from the user standpoint, the user didn't set it explicitly.

Reply:
we need to test it out.. I don't want to cause additional recompilation.

Reply:
Since it's not on the XLATensor, the replicated sharding for implicit replication doesn't get attached to the HLO, so it's different to the compiler as well. After this change, we'll move the sharding from the PjRtShardedData to the XLATensor, which will then be added to the HLO.

IMO it would be cleaner to not attach replication to the PjRtShardedData. I'll try this out and open a PR later today if there's no blockers.

Reply:
I think it does show up in the HLO for implicit replications?

Reply:
Hmm, maybe it's already being set from output sharding propagation? In that case, this wouldn't be an issue, we'd already have two compilations from the sharding propagation. I'll do a quick test to verify.

Reply:
The two compilations may be very well caused by the optimizer not the output sharding propagation? Also, I don't think we ever actually update the graph when the hashing get updated?  (This seems a serious problem. lol)

Reply:
Jiewen is right, implicit replication is showing in the HLO, so my original point here is invalid. IIRC, we intentionally left the annotation off the XLATensor to keep it out of the HLO so that the compiler has more flexibility to decide the sharding, but maybe that's not an issue now that output sharding propagation is enabled.

Let's confirm the desired behavior with Yeounoh when he's back.

Reply:
After the first step, all unsharded tensors should be Implicitly replicated. And in the subsequent steps, we would have cached the sharded computation and the shardings would have become stable. I think we could leave the replicated sharding spec on the tensors if they are implicitly replicated, since we decided to use PjRtShardedData for all tensors in SPMD mode. Before it wasn't the case, and we had to make the distinction. WDYT?